designs/documents.md

```diff
@@ -0,0 +1,424 @@
+date in a backwards-compatible way.
+
+### Serialization and Deserialization
+
+The major advantage of the `Document` type over just using `DocumentValue`
+directly is that it integrates with schema-based serialization and
+deserialization. Since `Document` implements `SerializeableShape` and
+`DeserializeableShape` it can be directly used with any `ShapeSerializer` or
+`ShapeDeserializer` without any additional effort. The same `Document` can be
+serialized to and from XML, JSON, CBOR, etc. without any loss of information.
+
+#### Protocol-Specific Documents
+
+Some media types may not support all of the Smithy data model. JSON, for
+example, does not have a native capability to store blobs. In current AWS
+protocols that use JSON, blobs are stored as base64-encoded strings. The base
+`Document` class is unaware of this, but a `JSONDocument` subclass can smooth
+over the issue by attempting to coerce a string into a blob when `as_blob`
+is called.
+
+```python
+class JSONDocument(Document):
+    @override
+    def as_blob(self) -> bytes:
+        if self.shape_type is ShapeType.STRING:
+            return b64decode(self.as_string())
+        return super().as_blob()
+```
+
+This same strategy can be used to support timestamps based on the
+`@timestampFormat`, to support representing `Decimal`s as strings to avoid
+precision loss, etc.
+
+Protocol-specific documents are also instrumental in fully supporting late-bound
+types.
+
+#### Late-Bound Types
+
+`Document`s may also be used to implement a system of late-bound types. That is,
+clients and servers may exchange `Document`s which represent particular modeled
+shapes. AWS services, for example, may send structured events to trigger a
+Lambda function. If the shape of that event is known, the `Document` may be
+converted to that shape.
+
+If the specific shape is known ahead of time with external knowledge, then
+`as_shape` can be used directly. But there are cases where the `Document` may
+represent one of several shapes. For this reason, `Document`s have a
+`discriminator` property that delclares what shape it represents. By default
+this is the id of the document's schema, but that schema itself will be a
+default schema from the prelude unless the document is specifically constructed
+with one.
+
+To reliably communicate the discriminator, it needs to be embedded in the
+document itself. How this is done is protocol-specific, and thus is the domain
+of protocol-specific documents. A JSON-based protocol could, for example,
+embed the shape id as the value of a top-level `__shape` property.
+
+```json
+{
+    "__shape": "smithy.example#ExampleStruct",
+    "foo": "spam",
+    "bar": "eggs
+}
+```
+
+A `TypeRegistry` would then be used to get the concrete class, which is then used
+to convert the `Document` into the target class.
+
+```python
+>>> document.as_shape(registry.get(document.discriminator))
+```
+
+##### Type Registries
```

Comment:
Design for these probably goes in the serialization document, but I wanted to have it here for now for the sake of making the review easier.